packages/core/nano/src/types.ts

```diff
@@ -1,22 +1,92 @@
-export interface Consumer<T> {
-    notify: <K extends keyof T>(property: K, value: T[K]) => void;
-    notifyMany: <K extends keyof T>(values: Pick<T, K>) => void;
+export interface Pending<T> {
+  kind: "Pending";
+  estimate?: T;
 }
 
-export interface Pending<T> {
-    kind: "Pending";
-    estimate?: T;
+/**
+ * The interface for an object whose data can be bound to. We use this contract for
+ * components that want to expose their data and its changes to other components.
+ *
+ * Any component that implements IProducer is expected to provide some registration
+ * functionality and to notify consumers whenever the data they are bound to changes.
+ */
+export interface IProducer<T> {
+  /**
+   * Unsubscribes a consumer from this producer.
+   * @param consumer - The consumer to unregister from the Producer.
+   */
+  removeConsumer(consumer: IConsumer<T>): void;
+
+  /**
+   * Returns a reader for this producer's values and implicitly subscribes the given
+   * consumer to change notifications from this producer (if it isn't already).
+   * 
+   * @param consumer - The object to be notified of value changes.
+   */
+  open(consumer: IConsumer<T>): {
```

Comment:
Is the invariant here that `prod.open(consumerA).read(p) === prod.open(consumerB).read(p)` if no updates have been sent? I guess it's likely to be the case that `prod.open(consumerA) === prod.open(consumerB)`, but I'm not sure we would need this as an invariant.

Reply:
The first invariant is effectively true.  If the two reads happen in separate turns, _technically_ one consumer could get a pending value while the second gets the final value, but both consumers will converge on the same value at the end of the recalc cycle.

Regarding if open() is required to return the same instance, I'd leave that up to the producer for now.

Reply:
Yep, that makes sense. I was just thinking about an even weaker invariant that only holds in the same turn. Basically a producer can't return a stateful reader, or return an observationally different reader to different consumers.